src/buildstream/_frontend/widget.py

```diff
@@ -437,6 +437,23 @@ def show_pipeline(self, dependencies, format_):
                 runtime_deps = [e._get_full_name() for e in element._dependencies(_Scope.RUN, recurse=False)]
                 line = p.fmt_subst(line, "runtime-deps", _yaml.roundtrip_dump_string(runtime_deps).rstrip("\n"))
 
+            # Source Information
+            if "%{source-info" in format_:
+
+                # Get all the SourceInfo objects
+                #
+                all_source_infos = []
+                for source in element.sources():
```

Comment:
A was playing with this a little, and came up with this:
```suggestion
                source_element = element._get_source_element()
                for source in source_element.sources():
```

given `Element._get_source_element()` gives the element that has the sources for this element, it really made sense. It gives a nice solution to the issue of actual sources of filter elements not being accessible.

I'm thinking we can push it further and move this whole loop into an `Element.collect_source_info()` method. This way, something like the `collect_manifest` plugin wouldn't need to call `Element.sources()` at all (and we can deprecate it). It would be able to call `_get_source_element()` as needed, and users won't have to worry about it.

Moving forward with this idea, we could even have the `compose` plugin collect sources from all build dependencies that contribute files to the final image. Though this might need some more thinking.

Reply:
We agreed to postpone this.